stairway-azure/src/main/java/bio/terra/stairway/azure/AzureServiceBusQueue.java

```diff
@@ -0,0 +1,240 @@
+package bio.terra.stairway.azure;
+
+import bio.terra.stairway.QueueInterface;
+import bio.terra.stairway.QueueProcessFunction;
+import bio.terra.stairway.exception.StairwayExecutionException;
+import com.azure.identity.DefaultAzureCredentialBuilder;
+import com.azure.messaging.servicebus.ServiceBusClientBuilder;
+import com.azure.messaging.servicebus.ServiceBusException;
+import com.azure.messaging.servicebus.ServiceBusMessage;
+import com.azure.messaging.servicebus.ServiceBusReceivedMessage;
+import com.azure.messaging.servicebus.ServiceBusReceiverClient;
+import com.azure.messaging.servicebus.ServiceBusSenderClient;
+import com.azure.messaging.servicebus.models.ServiceBusReceiveMode;
+import java.time.Duration;
+import java.util.List;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.commons.lang3.Validate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AzureServiceBusQueue implements QueueInterface {
+
+  private static final Logger logger = LoggerFactory.getLogger(AzureServiceBusQueue.class);
+
+  private final ServiceBusReceiverClient serviceBusReceiverClient;
+  private final ServiceBusSenderClient serviceBusSenderClient;
+
+  public static AzureServiceBusQueue.Builder newBuilder() {
+    return new AzureServiceBusQueue.Builder();
+  }
+
+  /**
+   * Construct the queue by creating a receiver and sender client for reading and writing to the
+   * subscription and reading from the topic.
+   *
+   * @param serviceBusReceiverClient the receiver client
+   * @param serviceBusSenderClient the sender client
+   */
+  AzureServiceBusQueue(
+      ServiceBusReceiverClient serviceBusReceiverClient,
+      ServiceBusSenderClient serviceBusSenderClient) {
+    this.serviceBusReceiverClient = serviceBusReceiverClient;
+    this.serviceBusSenderClient = serviceBusSenderClient;
+  }
+
+  /**
+   * Construct the queue by creating a receiver and sender client using the builder parameters.
+   *
+   * @param builder the builder used to pass parameters
+   */
+  public AzureServiceBusQueue(AzureServiceBusQueue.Builder builder) {
+    ServiceBusClientBuilder serviceBusClientBuilder;
+
+    if (builder.useManagedIdentity) {
+      serviceBusClientBuilder =
+          new ServiceBusClientBuilder()
+              .fullyQualifiedNamespace(builder.namespace)
+              .credential(new DefaultAzureCredentialBuilder().build());
+    } else {
+      serviceBusClientBuilder =
+          new ServiceBusClientBuilder().connectionString(builder.connectionString);
+    }
+
+    serviceBusReceiverClient =
+        serviceBusClientBuilder
+            .receiver()
+            .topicName(builder.topicName)
+            .subscriptionName(builder.subscriptionName)
+            .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
+            .disableAutoComplete()
+            .maxAutoLockRenewDuration(builder.maxAutoLockRenewDuration)
+            .buildClient();
+
+    serviceBusSenderClient =
```

Comment:
Do we want a `shutdown` method to close these clients?

Reply:
Yeah, I wondered about this as well. The GCP implementation has a shutdown method, however it is not part of the interface. Now that you mention it and just to be safe, I'll add the method for consistency's sake.